be/src/exec/vectorized/chunk_sorter_heapsorter.h

```diff
@@ -0,0 +1,263 @@
+        update_top();
+    }
+
+    void remove_top() {
+        std::pop_heap(_queue.begin(), _queue.end(), _comp);
+        _queue.pop_back();
+    }
+
+    void push(T&& rowcur) {
+        _queue.emplace_back(std::move(rowcur));
+        std::push_heap(_queue.begin(), _queue.end(), _comp);
+    }
+
+    Sequence&& sorted_seq() {
+        std::sort_heap(_queue.begin(), _queue.end(), _comp);
+        return std::move(_queue);
+    }
+
+    Sequence& container() { return _queue; }
+
+    // replace top if val less than top()
+    void replace_top_if_less(T&& val) {
+        if (_comp(val, top())) {
+            replace_top(std::move(val));
+        }
+    }
+
+private:
+    Sequence _queue;
+    Compare _comp;
+
+    size_t _next_child_index() {
+        size_t next_idx = 0;
+        if (next_idx == 0) {
+            next_idx = 1;
+            if (_queue.size() > 2 && _comp(_queue[1], _queue[2])) ++next_idx;
+        }
+        return next_idx;
+    }
+
+    void update_top() {
+        size_t size = _queue.size();
+        if (size < 2) return;
+
+        auto begin = _queue.begin();
+
+        size_t child_idx = _next_child_index();
+        auto child_it = begin + child_idx;
+
+        /// Check if we are in order.
+        if (_comp(*child_it, *begin)) return;
+
+        auto curr_it = begin;
+        auto top(std::move(*begin));
+        do {
+            /// We are not in heap-order, swap the parent with it's largest child.
+            *curr_it = std::move(*child_it);
+            curr_it = child_it;
+
+            // recompute the child based off of the updated parent
+            child_idx = 2 * child_idx + 1;
+
+            if (child_idx >= size) break;
+
+            child_it = begin + child_idx;
+
+            if ((child_idx + 1) < size && _comp(*child_it, *(child_it + 1))) {
+                /// Right child exists and is greater than left child.
+                ++child_it;
+                ++child_idx;
+            }
+
+            /// Check if we are in order.
+        } while (!(_comp(*child_it, top)));
+        *curr_it = std::move(top);
+    }
+};
+
+struct ChunkCursorComparator {
+    ChunkCursorComparator(const int* reverse, const int* nan_direction_hint)
+            : reverse(reverse), nan_direction_hint(nan_direction_hint) {}
+
+    bool operator()(const ChunkRowCursor& lhs, const ChunkRowCursor& rhs) const {
+        size_t l_row_id = lhs.row_id();
+        size_t r_row_id = rhs.row_id();
+        int order_by_columns_sz = lhs.data_segment()->order_by_columns.size();
+        for (int i = 0; i < order_by_columns_sz; ++i) {
+            int res = lhs.data_segment()->order_by_columns[i]->compare_at(
+                    l_row_id, r_row_id, *rhs.data_segment()->order_by_columns[i].get(), nan_direction_hint[i]);
+            if (res != 0) return res * reverse[i] < 0;
+        }
+
+        return false;
+    }
+
+private:
+    // 1 or -1
+    const int* reverse;
+    const int* nan_direction_hint;
```

Comment:
Why not use vector?

Reply:
In some tests with single column sorting (without Filter), I found that the performance of using std::vector was less than ideal, for reasons I haven't fully analyzed yet